### PR TITLE
Added checking tx-type using transactions permission contract for miners

### PR DESCRIPTION
**Pull request by VladLupashevskyi**
In this commit you can find solution to the issue [#7350](https://github.com/paritytech/parity/issues/7350) which I opened recently.

Now block miners are checking whether the tx type is allowed for sender using transaction permission contract.

### CODE REVIEW

**Comment by tomusdrw** [comment]
I think the check should actually be done **before** importing transactions to the queue as well. Otherwise we might be storing and propagating a lot of invalid transactions.

Should be done here:
https://github.com/paritytech/parity/blob/master/ethcore/src/miner/miner.rs#L675

Other than that: great work, thank you @VladLupashevskyi

File: ethcore/src/miner/miner.rs [hunk 1]
```diff
@@ -419,7 +420,15 @@ impl Miner {
 		for tx in transactions {
 			let hash = tx.hash();
 			let start = Instant::now();
-			let result = open_block.push_transaction(tx, None);
+			// Check whether transaction type is allowed for sender
+			let result = match self.engine.machine().verify_transaction(&tx, open_block.header(), chain.as_block_chain_client()) {
+				Err(Error::Transaction(TransactionError::NotAllowed)) => {
+					Err(From::from(TransactionError::NotAllowed))
```
Comment (rphmeier) [inline ethcore/src/miner/miner.rs]: favor `.into()` instead of `From::from`

File: ethcore/src/miner/miner.rs [hunk 2]
```diff
@@ -419,7 +420,15 @@ impl Miner {
 		for tx in transactions {
 			let hash = tx.hash();
 			let start = Instant::now();
-			let result = open_block.push_transaction(tx, None);
+			// Check whether transaction type is allowed for sender
+			let result = match self.engine.machine().verify_transaction(&tx, open_block.header(), chain.as_block_chain_client()) {
```
Comment (rphmeier) [inline ethcore/src/miner/miner.rs]: I guess `open_block`'s parent state always has to be available in the client, so this is fine.

**Comment by arkpar** [comment]
Agree with @tomusdrw. It is assumed that transaction queue contains valid transactions only.

**Comment by tomusdrw** [comment]
@arkpar although the state of the transaction might change in the meantime, with current queue we have no way of invalidating such transactions, so I'm ok with check in both places.

**Comment by arkpar** [comment]
Makes sense

File: ethcore/src/miner/miner.rs [hunk 3]
```diff
@@ -680,6 +698,12 @@ impl Miner {
 						Err(e)
 					},
 					Ok(transaction) => {
+						// This check goes here because verify_transaction takes SignedTransaction parameter
+						if self.engine.machine().verify_transaction(&transaction, &best_block_header, client.as_block_chain_client()).is_err() {
```
Comment (arkpar) [inline ethcore/src/miner/miner.rs]: The error here should not be interpreted as permission failure. Other checks might be added in the future. Best would just to pass it to the caller:
```rust
self.engine.machine().verify_transaction(&transaction, &best_block_header, client.as_block_chain_client())?;
```

**Comment by VladLupashevskyi** [comment]
@arkpar Agreed. And thank you, my Rust knowledge is much better now :) I didn't know about this amazing "?" operator before.

**Pull request merged**